testing: ErrorCheck should only call if err != nil

## Big Question

Should the `ErrorCheck` function be called even if there is no error? I could see it going either way.

* Provides ultimate flexibility to do expected error type things (perhaps deprecate `ExpectError`?)
* But, somewhat counterintuitive (must check `err != nil` before things like `err.Error()`)


### SDK version
<!---
Inspect your go.mod as below to find the version, and paste the result between the ``` marks below.

go mod edit -json | jq '.Require[] | select(.Path=="github.com/hashicorp/terraform-plugin-sdk/v2")'

If you are not running the latest version of the SDK, please try upgrading
because your bug may have already been fixed.

If the command above doesn't yield any results, it means you may either be using v1 of the SDK or
have not have migrated to the standalone SDK yet. See https://www.terraform.io/docs/extend/plugin-sdk.html
for more.
-->

```
2.3
```

### Relevant provider source code

<!--
Paste any Go code that you believe to be relevant to the bug
e.g. schema or implementation of CRUD for a given resource or data source
-->
```go
func testAccSkipErrorCheck(err error, t *testing.T) error {
	if err != nil { // <--- Should this be necessary?
		if strings.Contains(err.Error(), "Operations related to PublicDNS") {
			t.Skipf("skipping test; this partition (%s) does not support PublicDNS operations", testAccGetPartition())
		}

		if strings.Contains(err.Error(), "Regional control plane current does not support ") {
			t.Skipf("skipping test; this region (%s) does not support latency health check", testAccGetRegion())
		}
	}

	return err
}

...

func TestAccAWSRoute53DelegationSet_basic(t *testing.T) {
	rString := acctest.RandString(8)
	refName := fmt.Sprintf("tf_acc_%s", rString)
	resourceName := "aws_route53_delegation_set.test"

	resource.ParallelTest(t, resource.TestCase{
		PreCheck:        func() { testAccPreCheck(t) },
		ErrorCheck:      func(err error) error { return testAccSkipErrorCheck(err, t) },
		IDRefreshName:   resourceName,
		IDRefreshIgnore: []string{"reference_name"},
		Providers:       testAccProviders,
		CheckDestroy:    testAccCheckRoute53DelegationSetDestroy,
		Steps: []resource.TestStep{
			{
				Config: testAccRoute53DelegationSetConfig(refName),
				Check: resource.ComposeTestCheckFunc(
					testAccCheckRoute53DelegationSetExists(resourceName),
				),
			},
			{
				ResourceName:            resourceName,
				ImportState:             true,
				ImportStateVerify:       true,
				ImportStateVerifyIgnore: []string{"reference_name"},
			},
		},
	})
}
```

### Debug Output
<!--
Full debug output can be obtained by running Terraform with the environment variable `TF_LOG=trace`. Please create a GitHub Gist containing the debug output. Please do _not_ paste the debug output in the issue, since debug output is long.

Debug output may contain sensitive information. Please review it before posting publicly, and if you are concerned feel free to encrypt the files using the HashiCorp security public key.
-->

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x57eeca6]

goroutine 15 [running]:
testing.tRunner.func1.1(0x718ae40, 0xc629cc0)
	/usr/local/opt/go/libexec/src/testing/testing.go:1072 +0x30d
testing.tRunner.func1(0xc001308600)
	/usr/local/opt/go/libexec/src/testing/testing.go:1075 +0x41a
panic(0x718ae40, 0xc629cc0)
	/usr/local/opt/go/libexec/src/runtime/panic.go:969 +0x1b9
github.com/terraform-providers/terraform-provider-aws/aws.testAccSkipErrorCheck(0x0, 0x0, 0xc001308600, 0xc0013b75d8, 0xc0013b7618)
	/Volumes/develop/go/src/github.com/hashicorp/terraform-provider-aws/aws/provider_test.go:962 +0x26
github.com/terraform-providers/terraform-provider-aws/aws.TestAccAWSRoute53HealthCheck_Disabled.func2(0x0, 0x0, 0x0, 0xc0013873b0)
	/Volumes/develop/go/src/github.com/hashicorp/terraform-provider-aws/aws/resource_aws_route53_health_check_test.go:303 +0x3e
github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource.runNewTest(0x8da5ce0, 0xc001308600, 0x0, 0xc0013873b0, 0xc00136e390, 0x0, 0xc0013334d0, 0x0, 0x0, 0x837bd90, ...)
	/Volumes/develop/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/resource/testing_new.go:138 +0x5e7
github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource.Test(0x8da5ce0, 0xc001308600, 0x0, 0xc0013873b0, 0xc00136e390, 0x0, 0xc0013334d0, 0x0, 0x0, 0x837bd90, ...)
	/Volumes/develop/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/resource/testing.go:558 +0x3e5
github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource.ParallelTest(0x8da5ce0, 0xc001308600, 0x0, 0xc0013873b0, 0x0, 0x0, 0xc0013334d0, 0x0, 0x0, 0x837bd90, ...)
	/Volumes/develop/go/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.3.0/helper/resource/testing.go:496 +0xb8
github.com/terraform-providers/terraform-provider-aws/aws.TestAccAWSRoute53HealthCheck_Disabled(0xc001308600)
	/Volumes/develop/go/src/github.com/hashicorp/terraform-provider-aws/aws/resource_aws_route53_health_check_test.go:301 +0x6f8
testing.tRunner(0xc001308600, 0x8370ec0)
	/usr/local/opt/go/libexec/src/testing/testing.go:1123 +0xef
created by testing.(*T).Run
	/usr/local/opt/go/libexec/src/testing/testing.go:1168 +0x2b3
FAIL	github.com/terraform-providers/terraform-provider-aws/aws	16.793s
```

### Expected Behavior

Don't call `ErrorCheckFunc` if `err == nil`.

### Actual Behavior

Calls `ErrorCheckFunc` if `err == nil`.

### Steps to Reproduce

Run AWS provider test suite with code above.

### References
<!--
Are there any other GitHub issues (open or closed) or Pull Requests that should be linked here? For example:
-->
- #592 
- #568

--- a/helper/resource/testing_new.go
+++ b/helper/resource/testing_new.go
@@ -114,7 +114,7 @@ func runNewTest(t testing.T, c TestCase, helper *plugintest.Helper) {
 					t.Fatalf("Step %d/%d error running import, expected an error with pattern (%s), no match on: %s", i+1, len(c.Steps), step.ExpectError.String(), err)
 				}
 			} else {
-				if c.ErrorCheck != nil {
+				if err != nil && c.ErrorCheck != nil {
 					err = c.ErrorCheck(err)
 				}
 				if err != nil {
@@ -134,7 +134,7 @@ func runNewTest(t testing.T, c TestCase, helper *plugintest.Helper) {
 					t.Fatalf("Step %d/%d, expected an error with pattern, no match on: %s", i+1, len(c.Steps), err)
 				}
 			} else {
-				if c.ErrorCheck != nil {
+				if err != nil && c.ErrorCheck != nil {
 					err = c.ErrorCheck(err)
 				}
 				if err != nil {

